version/src/main/java/io/smallrye/common/version/VersionScheme.java

```diff
@@ -0,0 +1,134 @@
+     *         second version according to this version scheme
+     * @throws VersionSyntaxException if one or both of the versions have syntactic errors according to this scheme
+     */
+    int compare(String v1, String v2);
+
+    /**
+     * Determine if two versions are equal according to this version scheme.
+     *
+     * @param v1 the first version (must not be {@code null})
+     * @param v2 the second version (must not be {@code null})
+     * @return {@code true} if the versions are equal, or {@code false} otherwise
+     */
+    default boolean equals(String v1, String v2) {
+        return compare(v1, v2) == 0;
+    }
+
+    /**
+     * Get the canonical representation of this version, according to the current version scheme.
+     *
+     * @param original the possibly non-canonical version (must not be {@code null})
+     * @return the canonical representation of the version (not {@code null})
+     */
+    default String canonicalize(String original) {
+        return appendCanonicalized(new StringBuilder(original.length()), original).toString();
+    }
+
+    /**
+     * Append the canonical representation of this version to the given builder,
+     * according to the current version scheme.
+     *
+     * @param target the string builder to append to (must not be {@code null})
+     * @param original the possibly non-canonical version (must not be {@code null})
+     * @return the string builder that was passed in
+     */
+    StringBuilder appendCanonicalized(StringBuilder target, String original);
+
+    /**
+     * Validate the syntax of the given version.
+     *
+     * @param version the version to validate (must not be {@code null})
+     * @throws VersionSyntaxException if the syntax of the version is invalid according to this version scheme
+     */
+    default void validate(String version) throws VersionSyntaxException {
+        final VersionIterator itr = iterate(version);
+        while (itr.hasNext()) {
+            itr.next();
+        }
+    }
+
+    /**
+     * Iterate the canonicalized components of the given version according to the rules of this versioning scheme.
+     *
+     * @param version the version to iterate (must not be {@code null})
+     * @return the version iterator (not {@code null})
+     */
+    VersionIterator iterate(String version);
+
+    // @formatter:off
+    /**
+     * A basic versioning scheme which is roughly compatible with semantic versioning, GNU's scheme, the OpenSSL scheme,
+     * the numeric scheme used by Unix-style and ELF dynamic linkers, etc. The syntax of this scheme can be described
+     * with this grammar:
+     * <code><pre>
+
+version ::= underscore-sequence
+          | version "." underscore-sequence
+
+underscore-sequence ::= dash-sequence
+                      | underscore-sequence "_" dash-sequence
+
+dash-sequence ::= plus-sequence
+                | dash-sequence "-" plus-sequence
+
+plus-sequence ::= parts
+                | plus-sequence "+" parts
```

Comment:
Any particular reason to write these productions as left-recursive? It seems to me that

```
plus-sequence ::= parts
                | plus-sequence "+" parts
```

is equivalent to

```
plus-sequence ::= parts ("+" parts)*
```

And you already use `+` elsewhere in the grammar to mean repetition (with at least 1 occurence). So using `*` (to mean repetition, no occurrence required) should be fine IMHO -- both chars are quite conventional in grammars.

Reply:
No reason; I just like left-recursive grammars. :)